Add reference assembly for System.Private.CoreLib.dll

akoeplinger opened the pull request:
We now compile against the reference assembly in all places where we were compiling against the mono/coreclr System.Private.CoreLib.dll implementation assembly before.

The new reference assembly consumes sources from the existing contracts to avoid checking in a generated version of SPC.dll (this would add ~20k lines of .cs which is mostly duplicated with System.Runtime.cs)

~~The reference .cs file of contracts where all types are actually living in SPC like System.Runtime.cs or System.Numerics.Vectors.cs were moved into `src\libraries\System.Private.CoreLib\ref\contracts\` to keep them co-located  (@ericstj suggested this in a discussion we had, but happy to change it back if most people don't like it)~~.

Since a few contracts have only partially moved types to SPC we wrap contract types with `#if !BUILDING_CORELIB_REFERENCE` so we can hide them when compiling the SPC reference assembly. This needed a few GenAPI changes which are implemented here: https://github.com/dotnet/arcade/pull/10003.

Note that this means that the types which live in CoreLib are moved to the end of the ref .cs file which causes a GitHub diff to show up, but it is actually a no-op.

Regenerating the ref .cs files works the same way as before, by running the `GenerateReferenceAssemblySource` target in the contract's src\ folder.

Fixes https://github.com/dotnet/runtime/issues/67660

/cc @stephentoub @tannergooding @jkotas @vitek-karas @ericstj @ViktorHofer

akoeplinger committed:
Add reference assembly for System.Private.CoreLib.dll

akoeplinger commented on eng/resolveContract.targets:
```diff
@@ -60,11 +61,36 @@
   </Target>
-  
+
+  <Target Name="GenerateDocIdsForCoreLibOrUriProject"
```
I need to add a better name/comment for this:

What this does is it takes the SPC or Uri implementation assembly and generates a .txt with all public types in it.

That file is then later on passed to GenAPI as the list of types which should be "split out" with the conditional compilation logic.

jkotas commented:
> moved into src\libraries\System.Private.CoreLib\ref\contracts\ to keep them co-located (@ericstj suggested this in a discussion we had, but happy to change it back if most people don't like it).

Thoughts:
- It makes the contract and the tests non-collocated. I think I would have a mild preference for keeping the contracts in the same location as they are today.
- If we were to move them, do we need to `contracts` subdirectory - why not just `src\libraries\System.Private.CoreLib\ref`?

stephentoub commented:
> I think I would have a mild preference for keeping the contracts in the same location as they are today.

+1

ViktorHofer commented on src/libraries/sfx-src.proj:
```diff
@@ -14,8 +14,7 @@
                                    Microsoft.VisualBasic.Core\src\Microsoft.VisualBasic.Core.vbproj" />
-    <ProjectReference Include="@(AnyProject);
-                               $(CoreLibProject)"
+    <ProjectReference Include="@(AnyProject)"
```
Why did you change this? CoreLib is part of the sfx-src assemblies build.

akoeplinger replied:
the sfx trimming breaks in mysterious ways when I don't do this, I'll reach out to you on Teams to debug it :)

ViktorHofer commented on src/libraries/oob.proj:
```diff
@@ -8,6 +8,7 @@
 
   <ItemGroup Condition="'$(BuildTargetFramework)' == '$(NetCoreAppCurrent)' or '$(BuildTargetFramework)' == ''">
     <ProjectReference Include="sfx-src.proj" OutputItemType="SharedFrameworkAssembly" />
+    <ProjectReference Include="$(CoreLibProject)" OutputItemType="SharedFrameworkAssembly" />
```
This looks wrong. CoreLib isn't part of the out-of-band build.

akoeplinger replied:
see above comment about the sfx trimming

ViktorHofer commented on eng/references.targets:
```diff
@@ -94,6 +85,16 @@
 
+  <Target Name="ReplaceCoreLibSrcWithRefAssemblyForCompilation"
+          AfterTargets="FindReferenceAssembliesForReferences"
+          Condition="'@(_coreLibProjectReference)' != ''">
```
nit: this target should only run when not compiling against reference assemblies.

```suggestion
          Condition="'@(_coreLibProjectReference)' != '' and '$(CompileUsingReferenceAssemblies)' != 'true'">
```

ViktorHofer commented on eng/references.targets:
```diff
@@ -23,21 +23,12 @@
                                    '$(Configuration)' != '$(MonoConfiguration)'">Configuration=$(MonoConfiguration)</SetConfiguration>
     </ProjectReference>
-    <!-- If a CoreLib ProjectReference is present, make all P2P assets non transitive. -->
+    <!-- If a CoreLib src/ref ProjectReference is present, make all P2P assets non transitive. -->
```
nit: I don't think we should differentiate between the kind of ProjectReference, especially as we never directly P2P the ref CoreLib.

```suggestion
    <!-- If a CoreLib ProjectReference is present, make all P2P assets non transitive. -->
```

akoeplinger replied:
yup, that was a leftover from an earlier iteration

ViktorHofer commented:
As an alternative to manually creating the reference assembly and maintaining it, could we add a switch to Roslyn to not emit private fields into the reference assembly to get a stable MVID? We would only want to use such a switch for unshipping "private" reference assemblies like System.Private.CoreLib anyway. cc @jaredpar

ViktorHofer commented on src/libraries/shims/src/System.csproj:
```diff
@@ -7,4 +7,7 @@
   <ItemGroup>
     <Compile Include="System.forwards.cs" />
   </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="$(CoreLibProject)" />
+  </ItemGroup>
```
These shims projects which don't compile against references already had a reference to CoreLib before you moved CoreLib out of sfx-src.proj.

akoeplinger replied:
see above comment about the sfx trimming

akoeplinger commented on Directory.Build.props:
```diff
@@ -255,9 +255,12 @@
     <CoreLibProject Condition="'$(RuntimeFlavor)' == 'CoreCLR'">$([MSBuild]::NormalizePath('$(CoreClrProjectRoot)', 'System.Private.CoreLib', 'System.Private.CoreLib.csproj'))</CoreLibProject>
     <CoreLibProject Condition="'$(RuntimeFlavor)' == 'Mono'">$([MSBuild]::NormalizePath('$(MonoProjectRoot)', 'System.Private.CoreLib', 'System.Private.CoreLib.csproj'))</CoreLibProject>
     <CoreLibProject Condition="'$(UseNativeAotCoreLib)' == 'true'">$([MSBuild]::NormalizePath('$(CoreClrProjectRoot)', 'nativeaot', 'System.Private.CoreLib', 'src', 'System.Private.CoreLib.csproj'))</CoreLibProject>
+    <UriProject>$([MSBuild]::NormalizePath('$(LibrariesProjectRoot)', 'System.Private.Uri', 'src', 'System.Private.Uri.csproj'))</UriProject>
```
Types like `System.Uri` are exposed from the System.Runtime.dll reference assembly but live in System.Private.Uri.dll, we need to be able to hide those types when we include System.Runtime.cs in the SPC ref assembly (otherwise they'd conflict with the ones exposed from SPU). So it is basically the opposite case from how we need to hide types that live in contracts when including the .cs in the SPC ref.

akoeplinger committed:
PR feedback

akoeplinger committed:
Disable compiling against SPC ref assembly for NativeAOT System.Private.* assemblies

They access a lot of internals of SPC, not worth it right now.

akoeplinger committed:
Revert changes that turned out to not be necessary

ViktorHofer commented on src/coreclr/nativeaot/System.Private.DisabledReflection/src/System.Private.DisabledReflection.csproj:
```diff
@@ -1,4 +1,7 @@
 <Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <CompileUsingReferenceAssemblies>false</CompileUsingReferenceAssemblies>
+  </PropertyGroup>
```
Why is this change necessary?

ViktorHofer commented on eng/resolveContract.targets:
```diff
@@ -60,11 +61,38 @@
+      <GenAPIConditionalTypeListWrapOtherTypes Condition="'$(GenAPIConditionalTypeListWrapOtherTypes)' == ''">true</GenAPIConditionalTypeListWrapOtherTypes>
+    </PropertyGroup>
+    <ItemGroup>
+      <_genAPIResolvedCoreLibProjectReferenceDir Include="@(_ResolvedProjectReferencePaths->WithMetadataValue('MSBuildSourceProjectFile','$(CoreLibProject)')->'%(RootDir)%(Directory)')" />
```
This won't return anything if `ProjectForGenAPIDocIdGeneration` is not CoreLib and when CoreLib isn't referenced.

akoeplinger replied:
that's true, but that can't really happen since the target only runs when `$(ContractTypesPartiallyMovedToCoreLib)' == 'true' or '$(ContractTypesPartiallyMovedToUri)' == 'true'` (the first property implies you have a CoreLib reference and the second one only applies to Uri which also has one).

ViktorHofer replied:
Right, I understand that this works as-is today but it is fragile. Maybe you can come up with a better pattern to retrieve the corelib assembly or include an explicit error that checks that corelib is referenced and suggests to reference it if it isn't.

akoeplinger replied:
`@(ReferencePath->'%(RootDir)%(Directory)'->Distinct())` works as well (that's what the GenAPI targets from arcade does), so will switch to that

akoeplinger committed:
Apply suggestions from code review

Co-authored-by: Viktor Hofer <viktor.hofer@microsoft.com>

akoeplinger committed:
PR feedback

akoeplinger commented:
@jkotas I'm looking at it, this was mostly to get something building so I could get test coverage, I'll continue looking at how best to handle NativeAOT

akoeplinger committed:
Fix build of System.Diagnostics.Debug.Tests.csproj

akoeplinger committed:
Don't binplace System.Private.CoreLib ref assembly

akoeplinger committed:
Move contract .cs files back into the contract's ref\ directory

akoeplinger committed:
Add ApiCompat run for System.Private.CoreLib ref<->impl

akoeplinger committed:
Fix ApiCompat differences

akoeplinger commented:
@jkotas @stephentoub I've moved the contracts back to the original place.

We now also do an ApiCompat run to compare the SPC ref/impl, this will also catch issues with NativeAOT.

I think I'm pretty much done with what I had in mind so please feel free to review :)

ViktorHofer commented on src/libraries/apicompat/ApiCompat.proj:
```diff
@@ -95,6 +95,36 @@
+    </ItemGroup>
+  </Target>
+
+  <Target Name="RunApiCompatForCoreLib"
```
This projects compare the current live built assemblies against prebuilts (net6.0, netstandard2.1 and netstandard2.0). The ApiCompat comparison for CoreLib should just happen as part of the CoreLib build similar to how this works for every other library.

ViktorHofer commented on src/libraries/System.Private.CoreLib/ref/System.Private.CoreLib.csproj:
```diff
@@ -17,6 +17,10 @@
   </PropertyGroup>
+  <PropertyGroup>
+    <!-- disable binplacing since the System.Private.CoreLib reference assembly is internal to our build and shouldn't be exposed in ref packs -->
+    <DisableBinPlaceAssembly>true</DisableBinPlaceAssembly>
```
I'm surprised that ref/CoreLib is binplaced anywhere. CoreLib is already marked to not contribute a reference assembly to the targeting pack  / build: https://github.com/dotnet/runtime/blob/81a3dcdad5235c38a329b372af132a0053dc1708/src/libraries/NetCoreAppLibrary.props#L178.

If this doesn't work, there's probably a bug in that system that needs to be fixed.

akoeplinger replied:
As discussed on Teams, this is binplaced because of https://github.com/dotnet/runtime/blob/e48f21564bb04aca1599eb097d93cf95a8c2cbb1/src/libraries/Directory.Build.targets#L86-L91

ViktorHofer commented on src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/DiagnosticCounter.cs:
```diff
@@ -20,6 +20,11 @@ namespace System.Diagnostics.Tracing
     /// DiagnosticCounter is an abstract class that serves as the parent class for various Counter* classes,
     /// namely EventCounter, PollingCounter, IncrementingEventCounter, and IncrementingPollingCounter.
     /// </summary>
+#if !ES_BUILD_STANDALONE
+#if !FEATURE_WASM_PERFTRACING
+    [System.Runtime.Versioning.UnsupportedOSPlatform("browser")]
+#endif
+#endif
```
nit: this one and others:

```suggestion
#if !ES_BUILD_STANDALONE && !FEATURE_WASM_PERFTRACING
    [System.Runtime.Versioning.UnsupportedOSPlatform("browser")]
#endif
```

akoeplinger replied:
This was intentional to make it easier to grep for occurrences of `#if !FEATURE_WASM_PERFTRACING`

ViktorHofer replied:
Do we have any principle for this? I would prefer to not loose syntactic sugar for the sake of grepping.

ViktorHofer commented on src/libraries/System.Threading.Thread.WebAssembly.Threading/ref/System.Threading.Thread.WebAssembly.Threading.csproj:
```diff
@@ -13,7 +13,7 @@
   </PropertyGroup>
   <ItemGroup>
-    <Compile Include="$(LibrariesProjectRoot)System.Threading.Thread\ref\System.Threading.Thread.cs" />
+    <Compile Include="..\..\System.Threading.Thread\ref\System.Threading.Thread.cs" />
```
undo this change:

```suggestion
    <Compile Include="$(LibrariesProjectRoot)System.Threading.Thread\ref\System.Threading.Thread.cs" />
```

ViktorHofer commented on src/libraries/System.Threading.ThreadPool.WebAssembly.Threading/ref/System.Threading.ThreadPool.WebAssembly.Threading.csproj:
```diff
@@ -13,7 +13,7 @@
   </PropertyGroup>
   <ItemGroup>
-    <Compile Include="$(LibrariesProjectRoot)System.Threading.ThreadPool\ref\System.Threading.ThreadPool.cs" />
+    <Compile Include="..\..\System.Threading.ThreadPool\ref\System.Threading.ThreadPool.cs" />
```
undo this change:

```suggestion
    <Compile Include="$(LibrariesProjectRoot)System.Threading.ThreadPool\ref\System.Threading.ThreadPool.cs" />
```

ViktorHofer commented on src/libraries/System.Diagnostics.Debug/tests/System.Diagnostics.Debug.Tests.csproj:
```diff
@@ -1,10 +1,9 @@
-    <ResolveAssemblyWarnOrErrorOnTargetArchitectureMismatch>None</ResolveAssemblyWarnOrErrorOnTargetArchitectureMismatch>
     <TargetFrameworks>$(NetCoreAppCurrent)-windows;$(NetCoreAppCurrent)-Unix</TargetFrameworks>
     <TestRuntime>true</TestRuntime>
+    <CompileUsingReferenceAssemblies>false</CompileUsingReferenceAssemblies>
```
please add a comment with the reasoning

akoeplinger committed:
Better fix for wasm threading ApiCompat errors

akoeplinger committed:
Revert unnecessary changes

akoeplinger committed:
Add comment

akoeplinger committed:
Disable binplacing with existing property

akoeplinger committed:
Move CoreLib apicompat run to src projects

ViktorHofer commented on src/libraries/System.Diagnostics.Debug/tests/System.Diagnostics.Debug.Tests.csproj:
```diff
@@ -3,6 +3,7 @@
     <RootNamespace>System.Diagnostics.Tests</RootNamespace>
     <TargetFrameworks>$(NetCoreAppCurrent)-windows;$(NetCoreAppCurrent)-Unix</TargetFrameworks>
     <TestRuntime>true</TestRuntime>
+    <!-- some tests need types like System.Diagnostics.DebugProvider which are only exposed from System.Private.CoreLib -->
```
```suggestion
    <!-- Some tests need types like System.Diagnostics.DebugProvider which are only exposed in the implementation assembly of System.Private.CoreLib -->
```

akoeplinger replied:
that's not correct, it is exposed from the reference assembly of System.Private.CoreLib, just not System.Runtime.

ViktorHofer commented on src/libraries/System.Runtime/src/System.Runtime.csproj:
```diff
@@ -3,6 +3,14 @@
   </PropertyGroup>
+  <PropertyGroup>
+    <!-- the System.Runtime project is special, we need to be able to hide the System.Private.Uri types for the generated ref .cs so we can include it in the System.Private.CoreLib ref .cs -->
+    <ContractTypesPartiallyMovedToUri>true</ContractTypesPartiallyMovedToUri>
```
Can this be just `ContractTypesPartiallyMoved` instead of separate properties for CoreLib and Uri? From what I can see, this property is just a feature switch (target condition) and isn't used to infer the project.

ViktorHofer commented on eng/resolveContract.targets:
```diff
@@ -60,11 +65,35 @@
+          Condition="'@(ProjectReference)' != '' and @(ProjectReference->AnyHaveMetadataValue('Identity', '$(ProjectForGenAPIDocIdGeneration)')) and ('$(ContractTypesPartiallyMovedToCoreLib)' == 'true' or '$(ContractTypesPartiallyMovedToUri)' == 'true')">
+    <PropertyGroup>
+      <!-- generate conditional compilation symbols to hide types implemented in contracts so we can include the .cs in the System.Private.CoreLib ref -->
+      <GenAPIConditionalTypeList>$(IntermediateOutputPath)GenAPIConditionalTypeList.txt</GenAPIConditionalTypeList>
```
Just double checking, what happens if the IntermediateOutputPath doesn't yet exist?

ViktorHofer commented on eng/resolveContract.targets:
```diff
@@ -60,11 +65,35 @@
           BeforeTargets="GenerateReferenceAssemblySource">
     <PropertyGroup>
-      <GenAPIFollowTypeForwards Condition="'%(ProjectReference.Identity)' == '$(CoreLibProject)'">true</GenAPIFollowTypeForwards>
+      <GenAPIFollowTypeForwards Condition="'%(ProjectReference.Identity)' == '$(CoreLibProject)' or '%(ProjectReference.Identity)' == '$(UriProject)'">true</GenAPIFollowTypeForwards>
```
Unsure if that would work, but could/should this be conditioned on the `ProjectForGenAPIDocIdGeneration` project?

akoeplinger replied:
Yes this case happens for System.Runtime.dll, it has type forwarders to both CoreLib and Uri and both need to be followed.

ViktorHofer commented on src/coreclr/nativeaot/System.Private.DisabledReflection/src/System.Private.DisabledReflection.csproj:
```diff
@@ -1,4 +1,7 @@
 <Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <CompileUsingReferenceAssemblies>false</CompileUsingReferenceAssemblies>
```
Even though you explained this in the commit message, please add a comment to the projects in which you added this.

ViktorHofer commented on src/libraries/System.Private.CoreLib/ref/System.Private.CoreLib.csproj:
```diff
@@ -0,0 +1,57 @@
+        Remove warning disable when nullable attributes are respected,
+        Type has no accessible constructors which use only CLS-compliant types -->
+    <NoWarn>$(NoWarn);0809;0618;CS8614;CS3015</NoWarn>
+    <TargetFramework>$(NetCoreAppCurrent)</TargetFramework>
```
nit: please put the target framework at the very top. We aren't yet consistent and this isn't enforced anywhere but it helps with readability.

ViktorHofer commented on src/libraries/System.Private.CoreLib/ref/System.Private.CoreLib.csproj:
```diff
@@ -0,0 +1,57 @@
+    <DefineConstants>$(DefineConstants);HIDE_CONTRACT_TYPES</DefineConstants>
+  </PropertyGroup>
+  <PropertyGroup>
+    <!-- disable binplacing since the System.Private.CoreLib reference assembly is internal to our build and shouldn't be exposed in ref packs -->
```
It wouldn't be exposed in the targeting pack but in the artifacts/bin/ref folder, so anywhere makes more sense here.

```suggestion
    <!-- Disable binplacing since the System.Private.CoreLib reference assembly is internal to our build and shouldn't be exposed anywhere. -->
```

jaredpar commented:
> As an alternative to manually creating the reference assembly and maintaining it, could we add a switch to Roslyn to not emit private fields into the reference assembly produced via RefOut, in order to get a stable MVID? 

For context: the compiler emits the minimum set of members today that produce a functionally equivalent reference assembly. I understand that the ref assembly generation tools used by runtime today omit more members. From a language standpoint though these assemblies are  incorrect. They're not guaranteed to be semantically equivalent to their non-ref counterparts. The differences between the compiler and runtime tool is pretty small and the cases where it matters are often pretty obscure. However they are real cases and we have to respect them in the compiler. The runtime team gets away with this because they are dealing with a constrained scenario (a single code base) and between the runtime and compiler team we ensure the gaps don't turn into real customer facing issues. But we cannot categorically change the compiler ref assemblies to match the runtime ones because it's not a constrained scenario.

It's possible we could consider a feature flag for this behavior though. Essentially a `/features:ref-assembly-no-private` which is a "buyer beware" setting. I imagine it wouldn't be that hard to implement / maintain. The timeframe would be post .NET 7 though.

akoeplinger committed:
Fix ApiCompat issues by adding baseline files

akoeplinger committed:
PR feedback

marek-safar commented on src/mono/System.Private.CoreLib/MatchingRefApiCompatBaseline.txt:
```diff
@@ -0,0 +1,29 @@
+MembersMustExist : Member 'public void System.AppContext.remove_FirstChanceException(System.EventHandler<System.Runtime.ExceptionServices.FirstChanceExceptionEventArgs>)' does not exist in the reference but it does exist in the implementation.
+MembersMustExist : Member 'public void System.AppContext.remove_ProcessExit(System.EventHandler)' does not exist in the reference but it does exist in the implementation.
+MembersMustExist : Member 'public void System.AppContext.remove_UnhandledException(System.UnhandledExceptionEventHandler)' does not exist in the reference but it does exist in the implementation.
+MembersMustExist : Member 'public void System.Environment.FailFast(System.String, System.Exception, System.String)' does not exist in the reference but it does exist in the implementation.
```
Does this mean members like this should be private/internal?

marek-safar commented on src/mono/System.Private.CoreLib/MatchingRefApiCompatBaseline.txt:
```diff
@@ -0,0 +1,29 @@
+MembersMustExist : Member 'public void System.Environment.FailFast(System.String, System.Exception, System.String)' does not exist in the reference but it does exist in the implementation.
+MembersMustExist : Member 'protected void System.Diagnostics.Tracing.EventCounter.Flush()' does not exist in the reference but it does exist in the implementation.
+CannotMakeTypeAbstract : Type 'System.Diagnostics.Tracing.EventListener' is abstract in the reference but is not abstract in the implementation.
+CannotMakeMoreVisible : Visibility of member 'System.Diagnostics.Tracing.EventListener..ctor()' is 'protected' in the reference but 'public' in the implementation.
```
Why is that?

akoeplinger replied:
Not sure, but it is a pre-existing issue, see src/libraries/System.Diagnostics.Tracing/src/MatchingRefApiCompatBaseline.txt

akoeplinger replied:
@ViktorHofer I just tried but since the three CoreLib implementations expose different/additional public APIs in a few places right now it causes quite a mess with the "invalid/Unused baseline difference" logic in ApiCompat that would need conditionally including differerent MatchingRefApiCompatBaseline.txt which gets pretty hard to understand, I don't think it makes sense. Long term we should fix these inconsistencies and then we can do it.

ViktorHofer replied:
Can you please file follow-up issues to a) get the public apis in sync between corelibs and b) honor the MatchingRefApiCompatBaseline.txt files from the various libs?

akoeplinger replied:
Filed https://github.com/dotnet/runtime/issues/72594 and https://github.com/dotnet/runtime/issues/72595

akoeplinger commented on src/libraries/System.Runtime/src/System.Runtime.csproj:
```diff
@@ -3,6 +3,14 @@
+    <!-- the System.Runtime project is special, we need to be able to hide the System.Private.Uri types for the generated ref .cs so we can include it in the System.Private.CoreLib ref .cs -->
+    <ContractTypesPartiallyMoved>true</ContractTypesPartiallyMoved>
+    <ProjectForGenAPIDocIdGeneration>$(UriProject)</ProjectForGenAPIDocIdGeneration>
+    <GenAPIConditionalTypeListSymbol>!HIDE_URI_TYPES</GenAPIConditionalTypeListSymbol>
```
The way it works is that GenAPI will wrap the Uri types (which are implemented in System.Private.Uri.dll) with this conditional compilation symbol defined here so that we can exclude them when compiling the CoreLib reference assembly, e.g. 

```csharp
#if !HIDE_URI_TYPES
public class Uri
{
}
...
#endif
```

ericstj replied:
Why not just use the same define for this that you're using for others, EG: `BUILDING_REFERENCE`.  You could add an item with the typelist to apply it to the URI types.

Curious - does System.Private.Uri need a stable MVID too, or is that less of an issue since we don't cross-compile it as much?

akoeplinger replied:
@ericstj did that in the latest push. System.Private.Uri will have a stable MVID since it compiles against the CoreLib reference assembly too.

ViktorHofer commented:
@akoeplinger before merging this change in, please make sure that the GenAPI change in arcade is consumed so that we don't regress the dev inner loop.

akoeplinger commented:
@ViktorHofer sure. so we're good with the change in general? who can give an approval? 😄

ViktorHofer commented:
For the GenAPI change, I would ask @dotnet/area-infrastructure-libraries, @ericstj and @tannergooding to take a look. The change here looks good to me but I would like another pair of eyes to take a look.

steveisok commented:
@ericstj @tannergooding When you have a moment, can you please give this PR a good review?  It's an important change for Android and the quicker they can test it out, the better.

ericstj commented on eng/resolveContract.targets:
```diff
@@ -60,11 +65,36 @@
+    <PropertyGroup>
+      <!-- generate conditional compilation symbols to hide types implemented in contracts so we can include the .cs in the System.Private.CoreLib ref -->
+      <GenAPIConditionalTypeList>$(IntermediateOutputPath)GenAPIConditionalTypeList.txt</GenAPIConditionalTypeList>
+      <GenAPIConditionalTypeListSymbol Condition="'$(GenAPIConditionalTypeListSymbol)' == ''">!HIDE_CONTRACT_TYPES</GenAPIConditionalTypeListSymbol>
```
Have folks discussed the naming here?  This is confusing to me.  I'd prefer something like `BUILDING_REFERENCE` or something.  @stephentoub @tannergooding what is your preference since you're likely to be seeing this the most?

akoeplinger replied:
With the changes from https://github.com/dotnet/arcade/pull/10003 this was easier to do so I switched to using a single `BUILDING_CORELIB_REFERENCE` symbol.

akoeplinger committed:
Updates based on latest GenAPI changes

ericstj reviewed (commented):
GenAPI changes and usage look ok to me.  I’d like folks working in this codebase to provide their sign off.

akoeplinger committed:
Merge remote-tracking branch 'upstream/main' into corelib-ref-assembly

akoeplinger committed:
Update MatchingRefApiCompatBaseline.txt with recent changes

akoeplinger committed:
Update MatchingRefApiCompatBaseline.txt with recent changes

akoeplinger committed:
Fix MatchingRefApiCompatBaseline.txt validation failing on Windows

The `System.IO.Enumeration.FileSystemEntry.<_fileNameBuffer>e__FixedBuffer` only happens on non-Windows platforms.

ViktorHofer commented on src/mono/System.Private.CoreLib/System.Private.CoreLib.csproj:
```diff
@@ -361,6 +361,9 @@
     <ApiCompatValidateBaseline>true</ApiCompatValidateBaseline>
   </PropertyGroup>
+  <ItemGroup>
+    <MatchingRefApiCompatBaseline Condition="'$(TargetsWindows)' != 'true'" Include="MatchingRefApiCompatBaseline.Unix.txt" />
```
In the past, we never allowed API differences based on the platform that the assembly is targeting. Just wanted to bring this up.

Why not baseline the error for all platforms?

cc @ericstj

jkotas replied:
We have it in the shared baseline today: https://github.com/dotnet/runtime/blob/main/src/libraries/System.IO.FileSystem/src/MatchingRefApiCompatBaseline.txt#L8 . Would be consistent at least.

akoeplinger replied:
Yeah the problem is that this doesn't work in conjunction with the ApiCompatValidateBaseline setting (which ensures the baseline isn't containing unnecessary suppressions) since the `System.IO.Enumeration.FileSystemEntry.<_fileNameBuffer>e__FixedBuffer` only shows up on Unix platforms as it is the only one using the fixed array field: https://github.com/dotnet/runtime/blob/e415018a152a4a2e0db7619ceb91a3b3326b217a/src/libraries/System.Private.CoreLib/src/System/IO/Enumeration/FileSystemEntry.Unix.cs#L20

I talked to Roslyn and it's by design that the backing type is public even though the field is private (since the type is unspeakable so you can't actually do anything with it from C#) so I was planning on updating ApiCompat to ignore it.

The other option would be to turn off ApiCompatValidateBaseline.

jkotas replied:
https://github.com/dotnet/runtime/pull/73182 should make this problem go away.

akoeplinger replied:
Cool yeah that's a better option.

tannergooding reviewed (approved):
I didn't review in depth, but I don't see any changes that impact anything. It's just reordering where various entries are, so provided GenAPI changes are checked in it won't impact anyone needing to make subsequent changes to the ref assemblies.

akoeplinger committed:
Update GenAPI

akoeplinger committed:
Remove MatchingRefApiCompatBaseline.Unix.txt

No longer needed after https://github.com/dotnet/runtime/pull/73182

akoeplinger committed:
Merge remote-tracking branch 'upstream/main' into corelib-ref-assembly

jonathanpeppers commented:
@akoeplinger is there any plan to backport this to .NET 6?

We're hitting [this issue[(https://github.com/dotnet/linker/issues/2203) in MAUI + .NET 7 when you change the the `TargetFramework` to `net6.0-*`. So you end up with a .NET 7 linker and .NET 6 runtime packs in that case.

vitek-karas commented:
@sbomer - another instance of using 7.0 linker for 6.0 apps - completely different problem though.
I would argue this is even more of a case for using 6.0 linker for 6.0 apps regardless of the SDK... anyway :wink:

steveisok commented:
> @sbomer - another instance of using 7.0 linker for 6.0 apps - completely different problem though. I would argue this is even more of a case for using 6.0 linker for 6.0 apps regardless of the SDK... anyway 😉

@vitek-karas How realistic is making a change to use the 6.0 linker for 6.0 apps?  We're either going to have to do something like that or backport this change.  Using the 6.0 linker seems like the least impactful from my perspective.  Thoughts?

vitek-karas commented:
@sbomer has been looking into that, but it's relatively complicated and thus risky.

jonathanpeppers commented:
@sbomer so the result when building a Release-mode Android app, we end up with an assembly per architecture:

![image](https://user-images.githubusercontent.com/840039/188672577-a4ea396e-2979-4fce-bae0-afa6ebb293bb.png)
![image](https://user-images.githubusercontent.com/840039/188672588-04e67c30-f1bb-480c-8c2f-44ddb1bbe86b.png)

Where if you use a `global.json` and pin to the 6.0.400 SDK, you get 1 of each of these. We "dedup" them based on the MVID during the build.

I will get an app size comparison, but the duplicates cause another issue we need to solve first: https://github.com/xamarin/xamarin-android/issues/7302

I suspect the app might be near 4x larger?

marek-safar commented:
I think we ran out of time for any solution beyond the backport. Can we get that rolling ASAP?
